### PR TITLE
feat: add adapter for mento oracles

### PR DESCRIPTION
**Pull request by hatskier**

### CODE REVIEW

File: packages/on-chain-relayer/.env.mento-alfajores [hunk 1]
```diff
@@ -0,0 +1,14 @@
+RELAYER_ITERATION_INTERVAL=10000
```
Comment (cehali) [inline packages/on-chain-relayer/.env.mento-alfajores:1]: We already have folder with configs of deployed relayers: https://github.com/redstone-finance/redstone-oracles-monorepo/tree/main/packages/on-chain-relayer/deployed-config

File: packages/on-chain-relayer/test/relayer/update-prices.test.ts [hunk 2]
```diff
@@ -10,44 +10,98 @@ import {
...
   });
 
-  beforeEach(async () => {
-    const MangerContractFactory = await ethers.getContractFactory(
-      "PriceFeedsAdapterMock"
-    );
-    managerContract = await MangerContractFactory.deploy(dataFeedsIds);
-    await managerContract.deployed();
-  });
+  beforeEach(async () => {});
```
Comment (cehali) [inline packages/on-chain-relayer/test/relayer/update-prices.test.ts:27]: remove unnecessary

File: packages/on-chain-relayer/contracts/mocks/MentoAdapterMock.sol [hunk 3]
```diff
@@ -0,0 +1,20 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity ^0.8.4;
+
+import "@redstone-finance/evm-connector/contracts/mocks/AuthorisedMockSignersBase.sol";
+import "../custom-integrations/mento/MentoAdapter.sol";
+
+contract MentoAdapterMock is MentoAdapter, AuthorisedMockSignersBase {
```
Comment (cehali) [inline packages/on-chain-relayer/contracts/mocks/MentoAdapterMock.sol:8]: I wondering if we need this mock. If we want to mock only function maybe we could go with https://ethereum-waffle.readthedocs.io/en/latest/mock-contract.html?highlight=mock

File: packages/on-chain-relayer/src/core/contract-interactions/update-prices.ts [hunk 4]
```diff
@@ -1,38 +1,99 @@
...
-) => {
+): Promise<void> => {
   const dataPackagesTimestamps = Object.values(dataPackages).flatMap(
     (dataPackages) =>
       dataPackages.map(
         (dataPackage) => dataPackage.dataPackage.timestampMilliseconds
       )
   );
   const minimalTimestamp = Math.min(...dataPackagesTimestamps);
+  const wrappedAdapterContract =
```
Comment (cehali) [inline packages/on-chain-relayer/src/core/contract-interactions/update-prices.ts:35]: Why did you move the wrapping before the condition?

File: packages/on-chain-relayer/contracts/core/PermissionlessPriceUpdater.sol [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+    if (proposedTimestamp <= getLastUpdateTimestamp()) {
+      revert ProposedTimestampMustBeNewerThanLastTimestamp(
+        proposedTimestamp,
+        getLastUpdateTimestamp()
+      );
+    }
+  }
+
+  function isProposedRoundValid(uint256 proposedRound) private view returns (bool) {
+    return proposedRound == getLastRound() + 1;
```
Comment (jakub-wojciechowski) [inline packages/on-chain-relayer/contracts/core/PermissionlessPriceUpdater.sol:80]: Hmm, I see some benefits by iterating the round, but does the additional cost justify it. 
The efficiency of relayer is really crucial. 
Shouldn't the timestamps be enough to avoid race conditions (updating with the same timestamp)?

File: packages/on-chain-relayer/contracts/core/PermissionlessPriceUpdater.sol [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+    uint256 proposedRound,
+    uint256 proposedTimestamp
+  ) internal {
+    validateProposedRound(proposedRound);
+    validateProposedTimestamp(proposedTimestamp);
+    setLastRound(proposedRound);
+    setLastUpdateTimestamp(proposedTimestamp);
+  }
+
+  function validateDataPackageTimestampAgainstProposedTimestamp(
```
Comment (jakub-wojciechowski) [inline packages/on-chain-relayer/contracts/core/PermissionlessPriceUpdater.sol:39]: Do we allow receivedTimestampMilliseconds == lastUpdateTimestampMilliseconds?

File: packages/on-chain-relayer/contracts/custom-integrations/mento/MentoAdapter.sol [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+   * @param locationsInSortedLinkedLists The array of locations in linked list for reported values
+   */
+  function updatePriceValues(
+    uint256 proposedRound,
+    uint256 proposedTimestamp,
+    LocationInSortedLinkedList[] calldata locationsInSortedLinkedLists
+  ) public {
+    validateAndUpdateProposedRoundAndTimestamp(proposedRound, proposedTimestamp);
+
+    uint256 dataFeedsCount = getDataFeedsCount();
```
Comment (jakub-wojciechowski) [inline packages/on-chain-relayer/contracts/custom-integrations/mento/MentoAdapter.sol:104]: Do we really need to access storage and get tdataFeeds and count?
(It's expensive) + there might be inconsistency between dataFeeds from contract and the locationsInSortedLinkedList. Maybe it'll be cheaper/easier to pass both from relayer?

File: packages/on-chain-relayer/contracts/core/PermissionlessPriceUpdater.sol (hunk 6 above)
Comment (hatskier) [reply packages/on-chain-relayer/contracts/core/PermissionlessPriceUpdater.sol:39]: Yes

File: packages/on-chain-relayer/contracts/core/PermissionlessPriceUpdater.sol (hunk 5 above)
Comment (hatskier) [reply packages/on-chain-relayer/contracts/core/PermissionlessPriceUpdater.sol:80]: We will remove rounds in a separate PR

File: packages/on-chain-relayer/contracts/mocks/MentoAdapterMock.sol (hunk 3 above)
Comment (hatskier) [reply packages/on-chain-relayer/contracts/mocks/MentoAdapterMock.sol:8]: I did it in a consistent way with the price feed adapter ([PriceFeedsAdapterMock.sol](https://github.com/redstone-finance/redstone-oracles-monorepo/blob/8a84b787f82781740b1b4e51cabc7c32cadbcaab/packages/on-chain-relayer/contracts/mocks/PriceFeedsAdapterMock.sol)). It's also consistent with approach that we use for testing smart contracts in evm-connector. I've added a new issue to consider switching to waffle mock instead of mock contract files: https://github.com/redstone-finance/redstone-oracles-monorepo/issues/482

File: packages/on-chain-relayer/.env.mento-alfajores (hunk 1 above)
Comment (hatskier) [reply packages/on-chain-relayer/.env.mento-alfajores:1]: Removed

File: packages/on-chain-relayer/test/relayer/update-prices.test.ts (hunk 2 above)
Comment (hatskier) [reply packages/on-chain-relayer/test/relayer/update-prices.test.ts:27]: Removed

File: packages/on-chain-relayer/src/core/contract-interactions/update-prices.ts (hunk 4 above)
Comment (hatskier) [reply packages/on-chain-relayer/src/core/contract-interactions/update-prices.ts:35]: Moved after the condition

**Review by cehali — APPROVED** [review]
LGTM
